image, ipq40xx: fix typo, tftp installation, improve UBI safety checks, logging

guidosarducci opened the pull request:
### Description:

This series follows up from my earlier Linksys PR https://github.com/openwrt/openwrt/pull/11405 to address some observations and reported problems.

1. Fix a Linksys device typo in IMAGE_SIZE (75776K->75776k) that prevents parsing its value.
2. Add a helper to simplify expanding size units (k,m,g as above) and update existing code to use it.
3. Clean up Linksys footer creation: migrate the over-complex, convoluted `scripts/linksys-image.sh` to the existing make files as a simple build recipe.
4. Fix image size alignment needed by some U-Boot's TFTP image loading, as [previously](https://github.com/openwrt/openwrt/pull/11405#issuecomment-1358510123) [reported](https://github.com/openwrt/openwrt/pull/11405#issuecomment-1587517739) in the above PR. Do this by pre-padding the footer to ensure a pagesize-aligned final image size.
5. Support size-checks for NAND flashed factory images that account for UBI reserved PEBs. Add new device variable `NAND_SIZE` to support this.
6. Enable missing factory-image size checks for Linksys EA6350v3, EA8300, MR8300, WHW01 and WHW03v2, by adding the correct NAND_SIZE variables.
7. Revert a previous patch that breaks error messages printed during verbose builds.
8. Ensure size-check failures are indicated in the build system and not ignored as previously.


### Testing:

1. Tested OpenWrt factory image installation from both Linksys OEM GUI and OpenWrt Luci Sysupgrade, with Linksys EA8300.
2. Inspected OpenWrt factory image to confirm image size, footer placement and pre-padding are correct and also match Linksys OEM FW images.
3. Tested that build system raises errors for factory images that don't allow the required space for UBI bad-block reserves. (_NOTE: these images would pass the usual `check-size` rule._)
4. Verified upstream MTD/UBI documentation on NAND overhead by using `nandsim` and simulating OpenWrt image construction.


### Review Requests/Stakeholders:

@Borromini and @Nyshan previously reported the TFTP installation problems on their devices. I'm confident of the changes but if you have a chance to test TFTP installation I'd be grateful.

@robimarko @Ansuel @hauke @jeffsf  @mpratt14 did prior review, merge or related work.

@maxdad @badulesia @cybrnook @bartleyo and others helped with testing devices for the previous PR.

@aparcar authored the reverted code dealing with error messages and may have comments.

<br/><br/>

**Thanks to all for your feedback!**

guidosarducci committed:
ipq40xx: fix typo in Linksys WHW01 image definition

Use lower-case "k" in IMAGE_SIZE for Linksys WHW01, permitting proper unit
conversions in build recipes (e.g. 75776k -> 75776*1024).

Fixes: 2a9f3b7717 ("ipq40xx: fix up Linksys WHW01 board name, device definition")
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

DragonBluep commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -682,7 +682,7 @@ define Device/linksys_ea6350v3
 	UBINIZE_OPTS := -E 5
 	IMAGES += factory.bin
-	IMAGE/factory.bin := append-kernel | append-uImage-fakehdr filesystem | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=EA6350v3
+	IMAGE/factory.bin := append-kernel | append-uImage-fakehdr filesystem | pad-to $$$$(KERNEL_SIZE) | append-ubi | check-size | linksys-image type=EA6350v3
```
It doesn't make sense to check the size of the ubi image as ubifs needs 20 reserved PEBs to handle the NAND bad block.

guidosarducci replied:
Could you elaborate? How is the bad-block handling related to checking whether a user is building a kernel+filesystem image too large for their device? How then would such a mistake get caught or play out in practice?

robimarko replied:
Because 20 blocks are reserved by UBI for bad block handling by default so you have 20 blocks less of free space on every UBI volume.

guidosarducci replied:
Will comment in the main thread since I'm about to push some large changes, but I've lately dropped this "check-size".

guidosarducci committed:
image: use helper function for size units

Add the make function 'exp_units' for helping evaluate k/m/g size units in
expressions, and use this to consistently replace many ad hoc substitutions
like '$(subst k,* 1024,$(subst m, * 1024k,$(IMAGE_SIZE)))' in makefiles.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
image: improve UBI image sizing on NAND devices

Many NAND devices use a build recipe with "append-ubi | check-size" to
ensure factory images don't exceed the target flash partition size.
However, UBI reserves space for bad block handling and other operational
overhead, and thus 'check-size' can overestimate the space available by
several MB. In practice, this means a failed check is definitely a failure,
while a passing check is only probably a pass.

Improve the situation by teaching 'Build/append-ubi' to check image sizes
while accounting for UBI reserved blocks. Add new device variable NAND_SIZE
and use with existing IMAGE_SIZE to derate the available space. Each UBI
device reserves 20 PEBs per 1024 PEBs of the entire NAND device for bad
blocks, plus an additional 4 PEBs overhead.

Many devices can transparently enable this check by setting NAND_SIZE based
on their flash storage, and may then remove any unneeded 'check-size'.

Link: http://www.linux-mtd.infradead.org/doc/ubi.html#L_overhead

Suggested-by: Shiji Yang <yangshiji66@qq.com>
Suggested-by: Robert Marko <robimarko@gmail.com>
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
ipq40xx: enable UBI size checks for some Linksys NAND devices

Add correct NAND_SIZE in device definitions for EA6350v3, EA8300, MR8300,
WHW01 and WHW03v2, to enable improved image size checks wrt UBI reserved
blocks on NAND devices.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci committed:
image: fix Linksys image alignment and simplify footer creation

Current factory image sizes for Linksys devices are 256-byte aligned. This
is not an issue writing factory images from the OpenWrt or Linksys GUIs,
but can lead to failures using a TFTP client from the Linksys bootloader:

     NAND write: device 1 offset 0x2800000, size 0xc00100
     Attempt to write to non page aligned data
     NAND write to offset 2800000 failed -22
      0 bytes written: ERROR

Simplify Linksys footer creation by migrating to a makefile build recipe,
and pre-pad the footer (with 0xFF) to ensure the final image is $(PAGESIZE)
aligned.  Finally, remove the old linksys-image.sh script no longer needed.

Linksys footer details are given below for future reference. The 256-byte
footer is appended to factory images and tested by both the Linksys
Upgrader (observed in EA6350v3) and OpenWrt sysupgrade.

  Footer format:
    .LINKSYS.     Checked by Linksys upgrader before continuing.  (9 bytes)
    <VERSION>     Upgrade version number, unchecked so arbitrary. (8 bytes)
    <TYPE>        Model of device, space padded (0x20).          (15 bytes)
    <CRC>         CRC checksum of factory image to flash.         (8 bytes)
    <padding>     Padding ('0' + 0x20 * 7)                        (8 bytes)
    <signature>   Signature of signer, unchecked so arbitrary.   (16 bytes)
    <padding>     Padding with nulls (0x00)                     (192 bytes)

Link: https://github.com/openwrt/openwrt/pull/11405#issuecomment-1358510123
Link: https://github.com/openwrt/openwrt/pull/11405#issuecomment-1587517739

Reported-by: Stijn Segers <foss@volatilesystems.org>
Reported-by: Wyatt Martin <wawowl@gmail.com>
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

guidosarducci commented:
@robimarko @DragonBluep Thanks for your comments and forgive the long delay. This led me to a big detour into the UBI development history and documentation, and UBI-related building within OpenWrt. One thing to be aware of is that we have 40+ instances I've seen that include a build recipe with "append-ubi | check-size", so many see value in it. Practically speaking, a failed `check-size` is definitely a failure, but a passing `check-size` is only probably a pass, which is not ideal as you highlighted.

AIUI, UBI reserves PEBs for bad-block handling only on NAND, and the reserve depends on the size of the *entire NAND device*, as determined at first UBI attach and when the `root_data` volume autosizes. I believe I have a clean way to deal with this using a new `NAND_SIZE` device variable and will push an update shortly, so please have a look.

I've one other question for you regarding the other UBI space overhead: 2 PEB volume table, 1 PEB wear-leveling, 1 PEB atomic changes. Do you know if these are alreay included within the image we flash?

I believe the volume table is included (otherwise nothing works), but not sure of the other 2 PEBs, and would like to be accurate .

mpratt14 commented:
even if UBI size is dynamic we should be able to estimate what will be guaranteed to be able to fit on flash along with the extra space used for its functionality while running.

I would recommend a new build function like check-size-ubi which would do that, so it can be adjusted/fixed in one place instead of each instance it is used.

if there is truly no way to figure it out programmatically it can then just take some ratio of IMAGE_SIZE like maybe 0.9.

guidosarducci commented:
> even if UBI size is dynamic we should be able to estimate what will be guaranteed to be able to fit on flash along with the extra space used for its functionality while running.
> 
> I would recommend a new build function like check-size-ubi which would do that, so it can be adjusted/fixed in one place instead of each instance it is used.
> 
> if there is truly no way to figure it out programmatically it can then just take some ratio of IMAGE_SIZE like maybe 0.9
> 
> but maybe that should be saved for another PR

The current commit series actually handles all this, following several revisions over the last month. Size adjustment calculations are made and implemented, are all in one place, and in most cases only requires a device to set its `NAND_SIZE` variable. See https://github.com/guidosarducci/openwrt/commit/830954e21f76fb31b7da5ed49e2bf5982460871d and https://github.com/guidosarducci/openwrt/commit/bbc890fd9fb30352f34e29e9ff64fe07ede8f110 for an example of how simply this can work for some devices.

DragonBluep commented on include/image-commands.mk:
```diff
@@ -133,6 +133,8 @@ define Build/append-md5sum-ascii-salted
 	rm $@.salted
 endef
 
+UBI_NAND_SIZE_LIMIT= ($(IMAGE_SIZE) - ($(NAND_SIZE) * 20 / 1024))
```
The kernel config file says:
```
config MTD_UBI_BEB_LIMIT
	int "Maximum expected bad eraseblock count per 1024 eraseblocks"
```
So you can't use the nand flash size to calculate the reserved block size. Use attached mtd partition size instead.

guidosarducci replied:
No, it really does work that way. See for example: https://cateee.net/lkddb/web-lkddb/MTD_UBI_BEB_LIMIT.html
```
To put it differently, if this value is 20, UBI will try to reserve about 1.9% of physical
eraseblocks for bad blocks handling. And that will be 1.9% of eraseblocks on the entire
NAND chip, not just the MTD partition UBI attaches. This means that if you have, say,
a NAND flash chip admits maximum 40 bad eraseblocks, and it is split on two MTD
partitions of the same size, UBI will reserve 40 eraseblocks when attaching a partition.
```

I noticed that came as a surprise to many people on mailing lists during my research for this PR...

guidosarducci commented:
Rebased against master and updated the UBI NAND overhead after further review and testing. Hopefully that's it.

mpratt14 commented:
looking good, but a few comments

how positive are we that the new size check for UBI will be exact, or is it affected by the number of existing bad blocks? if so, maybe it would be safer to start with `IMAGE_SIZE * 15/16` so we can be absolutely sure nothing goes wrong. or maybe you can have 32 instead of 20 in that calculation to have a safer buffer of free space. (but comment that the real calculation is with 20)

perhaps instead of forcing lowercase k the new Make function "exp_units" can simply support uppercase instead? also, maybe the name of it can be "to-bytes"?

I don't think reverting verbose.mk is an appropriate solution, the "bad file descriptor" error has been a long standing problem in other cases, and I believe it will still exist after reverting that, for example, when running scripts/feeds. I believe it should be handled in a separate PR. I have a commit that might fix it from my old RFC PR that I can post later unless you want to pull it or try to fix it yourself.

guidosarducci commented:
> how positive are we that the new size check for UBI will be exact, or is it affected by the number of existing bad blocks? if so, maybe it would be safer to start with `IMAGE_SIZE * 15/16` so we can be absolutely sure nothing goes wrong. or maybe you can have 32 instead of 20 in that calculation to have a safer buffer of free space. (but comment that the real calculation is with 20)

Review the upstream UBI FAQ/documentation and you'll find that the objective of the UBI reserved space is to handle expected bad blocks over the device lifetime as specified by manufacturers, so creating new, ad-hoc heuristics as you suggest isn't needed.

> perhaps instead of forcing lowercase k the new Make function "exp_units" can simply support uppercase instead? also, maybe the name of it can be "to-bytes"?

Nothing is being forced; we just follow the existing case convention, and `exp_units` just centralizes the definition for better maintainability. I really don't want to mix in ideology or bike-shedding into things, just fix some problems and move forward. 😃 
 
> I don't think reverting verbose.mk is an appropriate solution, the "bad file descriptor" error has been a long standing problem in other cases, and I believe it will still exist after reverting that, for example, when running scripts/feeds. I believe it should be handled in a separate PR. I have a commit that might fix it from my old RFC PR that I can post later unless you want to pull it or try to fix it yourself.

As noted in my reversion, commit aee3594ffcb7 ("verbose.mk: print ERROR messages in non-verbose") breaks existing verbose mode logging for `check-size` (and the reasons are clear on inspecting the code), so it definitely needs rolling back. I've already tested that this restores verbose-mode logging. That commit can always be reworked and submitted again later, but we don't want to leave things broken as they are now.

I'll ping @aparcar for his thoughts and to revisit that commit. @mpratt14 Perhaps you can share your RFC PR too if it's related.

I'll rebase against master once again, but otherwise see no reason to delay merging this.

guidosarducci commented:
@robimarko @DragonBluep  FYI, I added your "Suggested-by:" to the commit accounting for NAND reserved blocks, if that's OK.

@hauke I'd appreciate if you would take a look at this series too.

mpratt14 commented:
> Nothing is being forced; we just follow the existing case convention, and `exp_units` just centralizes the definition for better maintainability. I really don't want to mix in ideology or bike-shedding into things, just fix some problems and move forward.

I don't mean to be dramatic, I was just thinking it would lessen the maintenance burden if we were to simply allow both upper and lower cases for units. Let's say this same typo happens again which doesn't even result in an error in regular builds, we would need yet another commit just for one character and that's _iff_ someone as detailed as you manages to catch it while reading makefiles... I say if we can easily program to turn future human mistakes into nothing, let's do so.

> That commit can always be reworked and submitted again later, but we don't want to leave things broken as they are now.

Agree, except that we can likely just fix the root cause of the problem which is deeper than a revert right now without waiting or reverting. I'm just going to post the commit I'm talking about in a new PR (sorry I wasn't able to earlier).

aparcar commented:
@guidosarducci please feel free to revert that commit. I'll see if I can find another way to fix the thing

mpratt14 commented:
@aparcar I believe I have a solution without having to revert it, I'm just doing a little more testing and then posting a PR soon, then we can decide which way to go

mpratt14 commented:
see #14039

guidosarducci commented:
> @guidosarducci please feel free to revert that commit. I'll see if I can find another way to fix the thing

@aparcar OK, thanks. I'll leave it in for now and assume any fix can be rebased on this.

@mpratt14 Had a quick look at the PR and left some comments.

I just rebased against master again to ease merging.

hauke commented on include/image-commands.mk:
```diff
@@ -223,6 +223,7 @@ define Build/check-size
 	[ $$limitsize -ge $$imagesize ] || { \
 		$(call ERROR_MESSAGE,    WARNING: Image file $@ is too big: $$imagesize > $$limitsize); \
 		rm -f $@; \
+		false; \
```
I think the build in the build bots is intentionally not failing when the size check fails, it just does not create the image.

robimarko replied:
I think so as well, as otherwise the whole target building would fail in buildbots and none of the other devices will get updated images.

guidosarducci replied:
> I think the build in the build bots is intentionally not failing when the size check fails, it just does not create the image.

The problem I've seen is that broken images **_are still created_** even if the size check fails, because later steps in the device build "pipeline" do so (e.g. footer creation). Since no one is checking for the related error messages, this would go unnoticed and allow publishing broken images.

Allowing builds to continue on size-check failures isn't really a solution; those failures are meant to be acted on. Failure of the whole target build might seem like a problem, but it means the device issue is noticed and fixed (marked broken, image updated).

Perhaps it's possible to "abort" just the device build on error so no image is created, while allowing the rest of the target build to continue. Is that possible? In any case, we need to handle these errors better...

mpratt14 replied:
instead of a `false` can it just be a `rm` so the bad image gets deleted but exit code remains good?

mpratt14 replied:
well wait a sec, the `rm` is right there so why isn't it getting removed?

mpratt14 replied:
this should be removing every possible thing that future build recipe steps rely on, maybe deleting the per device dtb as well will do it?

if that's too difficult, maybe create a stampfile that will be checked later as an independent make target in the rules for `make world`, in other words, create an actual make target called check-size, that gets called right before or after package/index is ran...

mpratt14 replied:
also a possibility, if we create a stampfile it can be checked for directly after the build recipe

we have some parts of image.mk where

`Build/check-size`
is called right after
`concat_cmd`
so we probably just need more of that...

hauke commented on include/verbose.mk:
```diff
@@ -29,15 +29,15 @@ ifeq ($(IS_TTY),1)
 
+  define ERROR_MESSAGE
+	printf "$(_R)%s$(_N)\n" "$(1)" >&8
+  endef
```
I think we should move this pat to #14039.

mpratt14 replied:
my PR is meant to solve that issue without needing this revert patch

guidosarducci replied:
Yes, I'll leave this for Michael and Paul to dive into. Main thing is that it does get resolved...

hauke commented:
Please remove the last 2 patches for now, the rest looks ok.

guidosarducci commented:
@hauke Thanks, I've dropped those commits and rebased, but please see my comment above regarding error-handling.

hauke commented:
Thank you for the PR, I merged it into main in c66511bc488c3665d47dd60baf981ca1c385c221, b16e14a220fcab19e3145328057e7188f00b6a17, cd5e0134b6bd2e34ad237d985da99c5307816c93, fc16df9fdd596d1a3f5c3be897dc0cf69c84c469 and a7818e0550a2594d50d2874c1321784be093d8d2.

guidosarducci commented:
Thanks, Hauke
